### PR TITLE
Add tangle walkers

### PR DESCRIPTION
**Pull request by thibault-martinez**
## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: bee-tangle/src/walkers/bfs.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        }
+    }
+}
+
+/// A walker that goes through the tangle in a Breadth First Search manner.
+pub struct TangleBfsWalker<'a> {
+    tangle: &'a Tangle,
+    parents: VecDeque<MessageId>,
+    visited: HashSet<MessageId>,
+    condition: Box<dyn Fn(&'a Tangle, &MessageId, &MessageData) -> bool>,
```
Comment (samuel-rufi) [inline bee-tangle/src/walkers/bfs.rs:50]: maybe rename this? as it does´t tell which condition this is? same for the DfsWalker.
maybe `match_cond` `?

File: bee-tangle/src/walkers/bfs.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    type Item = TangleWalkerItem;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let message_id = self.parents.pop_front()?;
+
+            if !self.visited.contains(&message_id) {
+                self.visited.insert(message_id);
+
+                return match self.tangle.get(&message_id) {
```
Comment (samuel-rufi) [inline bee-tangle/src/walkers/bfs.rs]: do we need this return here?

**Review by samuel-rufi — CHANGES_REQUESTED** [review]

File: bee-tangle/src/walkers/bfs.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+impl<'a> Iterator for TangleBfsWalker<'a> {
+    type Item = TangleWalkerItem;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let message_id = self.parents.pop_front()?;
+
+            if !self.visited.contains(&message_id) {
+                self.visited.insert(message_id);
```
Comment (Alex6323) [inline bee-tangle/src/walkers/bfs.rs]: `if self.visited.insert(message_id) { ... } `

File: bee-tangle/src/walkers/bfs.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                    Some(message_data) => {
+                        if (self.condition)(self.tangle, &message_data) {
+                            self.parents
+                                .extend(message_data.message().parents().iter().map(|p| p.id()));
+                            Some(TangleWalkerItem::Matched(message_id, message_data))
+                        } else {
+                            Some(TangleWalkerItem::Skipped(message_id, message_data))
+                        }
+                    }
+                    None => Some(TangleWalkerItem::Missing(message_id)),
```
Comment (Alex6323) [inline bee-tangle/src/walkers/bfs.rs]: consider using `if let Some(...) = self.tangle.get(&message_id) { ... } else { ... }`

File: bee-tangle/src/walkers/bfs.rs (hunk 2 above)
Comment (Alex6323) [reply bee-tangle/src/walkers/bfs.rs]: agree with Sam

Comment (thibault-martinez) [reply bee-tangle/src/walkers/bfs.rs]: Yes, we do

File: bee-tangle/src/walkers/dfs.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+impl<'a> Iterator for TangleDfsWalker<'a> {
+    type Item = TangleWalkerItem;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let message_id = self.parents.pop()?;
+
+            if !self.visited.contains(&message_id) {
+                self.visited.insert(message_id);
```
Comment (Alex6323) [inline bee-tangle/src/walkers/dfs.rs]: same as above (insert on a hashset returns a boolean, indicating whether the insert happened or not)

File: bee-tangle/src/walkers/dfs.rs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                        if (self.condition)(self.tangle, &message_id, &message_data) {
+                            self.parents
+                                .extend(message_data.message().parents().iter().rev().map(|p| p.id()));
+                            Some(TangleWalkerItem::Matched(message_id, message_data))
+                        } else {
+                            Some(TangleWalkerItem::Skipped(message_id, message_data))
+                        }
+                    }
+                    None => Some(TangleWalkerItem::Missing(message_id)),
+                };
```
Comment (Alex6323) [inline bee-tangle/src/walkers/dfs.rs]: * remove `return ... ;`
* consider using `if let`

File: bee-tangle/src/walkers/item.rs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::MessageData;
+
+use bee_message::MessageId;
+
+/// The type yielded by [`Tangle`](crate::Tangle) walkers.
```
Comment (Alex6323) [inline bee-tangle/src/walkers/item.rs:8]: I think you can remove the `(crate::Tangle)`. This should be resolved automatically for types within a crate.

Comment (thibault-martinez) [reply bee-tangle/src/walkers/item.rs:8]: No, it needs to be on the scope.

File: bee-tangle/src/walkers/bfs.rs (hunk 3 above)
Comment (thibault-martinez) [reply bee-tangle/src/walkers/bfs.rs]: Done

File: bee-tangle/src/walkers/dfs.rs (hunk 5 above)
Comment (thibault-martinez) [reply bee-tangle/src/walkers/dfs.rs]: Done

File: bee-tangle/src/walkers/bfs.rs (hunk 1 above)
Comment (thibault-martinez) [reply bee-tangle/src/walkers/bfs.rs:50]: To me, in the context of a walker, I feel like condition is enough as there is only one condition that can exist, whether to walk or not.

File: bee-tangle/src/walkers/bfs.rs (hunk 4 above)
Comment (thibault-martinez) [reply bee-tangle/src/walkers/bfs.rs]: I don't like the way the formatter renders it, it's way more confusing than the match but alright

File: bee-tangle/src/walkers/dfs.rs (hunk 6 above)
Comment (thibault-martinez) [reply bee-tangle/src/walkers/dfs.rs]: I did if let, the return is needed

**Review by Alex6323 — APPROVED** [review]

**Review by samuel-rufi — APPROVED** [review]

**Review by pvdrz — APPROVED** [review]

**Pull request merged**